### PR TITLE
Authentication to https nodes now possible

### PR DESCRIPTION
**Pull request by kuhlmannmarkus**
From here on the "host" has to be defined with the appropriate protocol: http:// or https://
When defining https, the user can now specifiy username and password for the authentication in the constructors for Core an Extended objects.
This resembles the "--remote-auth" parameter, which can be given to the iri.

### CODE REVIEW

**Comment by thibault-martinez** [comment]
Hi @kuhlmannmarkus, thanks for the PR !
This is introducing breaking change but I think you're right, it is necessary.

**Comment by thibault-martinez** [comment]
Please for further PRs use the provided .clang-format, CodeFactor won't annoy you anymore :)

**Comment by kuhlmannmarkus** [comment]
everything should be fine now

File: include/iota/api/service.hpp [hunk 1]
```diff
@@ -74,13 +75,17 @@ class Service {
...
     auto body    = cpr::Body{ data.dump() };
     auto headers = cpr::Header{ { "Content-Type", "application/json" },
                                 { "Content-Length", std::to_string(body.size()) },
                                 { "X-IOTA-API-Version", APIVersion } };
-    auto res     = cpr::Post(url, body, headers, cpr::Timeout{ timeout_ * 1000 });
 
+    auto res     = cpr::Post(url, body, headers, cpr::Timeout{ timeout_ * 1000 });;
+    if(user_.compare("") != 0 && pass_.compare("") != 0 && host_.compare(0,5,"https") == 0){
+      auto auth = cpr::Authentication{user_, pass_};
+      res = cpr::Post(url, body, headers, cpr::Timeout{ timeout_ * 1000 }, auth);
```
Comment (thibault-martinez) [inline include/iota/api/service.hpp]: You are still posting twice in case of an authenticated request

Comment (kuhlmannmarkus) [reply include/iota/api/service.hpp]: solved

File: include/iota/api/service.hpp [hunk 2]
```diff
@@ -74,13 +75,18 @@ class Service {
...
     auto body    = cpr::Body{ data.dump() };
     auto headers = cpr::Header{ { "Content-Type", "application/json" },
                                 { "Content-Length", std::to_string(body.size()) },
                                 { "X-IOTA-API-Version", APIVersion } };
-    auto res     = cpr::Post(url, body, headers, cpr::Timeout{ timeout_ * 1000 });
-
+    cpr::Response res;
+    if(user_.compare("") != 0 && pass_.compare("") != 0 && host_.compare(0,5,"https") == 0){
+      auto auth = cpr::Authentication{user_, pass_};
+      res = cpr::Post(url, body, headers, cpr::Timeout{ timeout_ * 1000 }, auth);
```
Comment (thibault-martinez) [inline include/iota/api/service.hpp]: You can even construct it in the parameters, as we do with cpr::Timeout

File: include/iota/api/service.hpp [hunk 3]
```diff
@@ -74,13 +75,18 @@ class Service {
...
-    auto url     = cpr::Url{ "http://" + host_ + ":" + std::to_string(port_) };
+    auto url     = cpr::Url{ host_ + ":" + std::to_string(port_) };
     auto body    = cpr::Body{ data.dump() };
     auto headers = cpr::Header{ { "Content-Type", "application/json" },
                                 { "Content-Length", std::to_string(body.size()) },
                                 { "X-IOTA-API-Version", APIVersion } };
-    auto res     = cpr::Post(url, body, headers, cpr::Timeout{ timeout_ * 1000 });
-
+    cpr::Response res;
+    if(user_.compare("") != 0 && pass_.compare("") != 0 && host_.compare(0,5,"https") == 0){
```
Comment (thibault-martinez) [inline include/iota/api/service.hpp]: Maybe use `!user_.empty()` and `!pass_.empty()` for clarity

**Comment by thibault-martinez** [comment]
That's it for my suggestions, maybe @Cylix wants to add something.
We'll wait CI reports too, thanks again !

**Comment by kuhlmannmarkus** [comment]
The errors shown by travis might indicate a problem with the IRI. I don't think, there is a problem with the code.

**Comment by thibault-martinez** [comment]
Definitely not your code, I'll investigate.

**Comment by Cylix** [comment]
Sorry for the delay, sounds good to me.
Will check for the failing build.

Thanks for your contribution!

**Comment by thibault-martinez** [comment]
Hi @kuhlmannmarkus , I'm waiting for CIs to finish on #305 and I'll merge this !

**Comment by thibault-martinez** [comment]
Thanks again @kuhlmannmarkus  !

**Pull request merged**